File: electron-frontend/src/main/updateChecker.ts
import { net } from 'electron'
import { configStore as store } from './configStore'

export interface AvailableUpdate {
  currentVersion: string
  latestVersion: string
  releaseUrl: string
  publishedAt: string | null
}

const RELEASE_API_URL = 'https://api.github.com/repos/NoGoat/Track-N-Race/releases/latest'
export const RELEASE_PAGE_URL = 'https://github.com/NoGoat/Track-N-Race/releases/latest'
const CHECK_INTERVAL_MS = 24 * 60 * 60 * 1000
const ENABLED_KEY = 'updates.enabled'
const LAST_CHECK_KEY = 'updates.lastCheckAt'
const SKIPPED_VERSION_KEY = 'updates.skippedVersion'

type ParsedVersion = {
  core: [number, number, number]
  prerelease: Array<number | string>
}

function parseVersion(value: string): ParsedVersion | null {
  const match = value.trim().match(/^v?(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$/)
  if (!match) return null

  return {
    core: [Number(match[1]), Number(match[2]), Number(match[3])],
    prerelease: match[4]
      ? match[4].split('.').map(part => /^\d+$/.test(part) ? Number(part) : part)
      : [],
  }
}

function compareVersions(leftValue: string, rightValue: string): number | null {
  const left = parseVersion(leftValue)
  const right = parseVersion(rightValue)
  if (!left || !right) return null

  for (let index = 0; index < left.core.length; index += 1) {
    if (left.core[index] !== right.core[index]) return left.core[index] > right.core[index] ? 1 : -1
  }

  if (left.prerelease.length === 0 && right.prerelease.length === 0) return 0
  if (left.prerelease.length === 0) return 1
  if (right.prerelease.length === 0) return -1

  const partCount = Math.max(left.prerelease.length, right.prerelease.length)
  for (let index = 0; index < partCount; index += 1) {
    const leftPart = left.prerelease[index]
    const rightPart = right.prerelease[index]
    if (leftPart === undefined) return -1
    if (rightPart === undefined) return 1
    if (leftPart === rightPart) continue
    if (typeof leftPart === 'number' && typeof rightPart === 'string') return -1
    if (typeof leftPart === 'string' && typeof rightPart === 'number') return 1
    return leftPart > rightPart ? 1 : -1
  }

  return 0
}

function normalizedVersion(value: string): string {
  return value.trim().replace(/^v(?=\d)/i, '')
}

function mockedUpdate(currentVersion: string): AvailableUpdate {
  return {
    currentVersion,
    latestVersion: '999.0.0',
    releaseUrl: RELEASE_PAGE_URL,
    publishedAt: new Date().toISOString(),
  }
}

export async function checkForUpdateOnStartup(currentVersion: string): Promise<AvailableUpdate | null> {
  if (process.env.TRACK_N_RACE_UPDATE_TEST === '1') return mockedUpdate(currentVersion)
  if (!(store.get(ENABLED_KEY, true) as boolean)) return null

  const now = Date.now()
  const lastCheckAt = store.get(LAST_CHECK_KEY, 0)
  if (typeof lastCheckAt === 'number' && now - lastCheckAt < CHECK_INTERVAL_MS) return null

  // Record the attempt before making the request so an outage cannot cause a
  // network request on every app launch during the next 24 hours.
  store.set(LAST_CHECK_KEY, now)

  try {
    const response = await net.fetch(RELEASE_API_URL, {
      headers: {
        Accept: 'application/vnd.github+json',
        'User-Agent': 'Track-N-Race',
        'X-GitHub-Api-Version': '2022-11-28',
      },
    })
    if (!response.ok) {
      console.warn(`[updates] GitHub returned HTTP ${response.status}`)
      return null
    }

    const body = await response.json() as Record<string, unknown>
    if (typeof body.tag_name !== 'string') {
      console.warn('[updates] latest release did not include a tag name')
      return null
    }

    const latestVersion = normalizedVersion(body.tag_name)
    const current = normalizedVersion(currentVersion)
    const comparison = compareVersions(latestVersion, current)
    if (comparison === null) {
      console.warn(`[updates] could not compare versions ${current} and ${latestVersion}`)
      return null
    }
    if (comparison <= 0) return null

    const skippedVersion = store.get(SKIPPED_VERSION_KEY, '')
    if (typeof skippedVersion === 'string' && normalizedVersion(skippedVersion) === latestVersion) return null

    return {
      currentVersion: current,
      latestVersion,
      releaseUrl: RELEASE_PAGE_URL,
      publishedAt: typeof body.published_at === 'string' ? body.published_at : null,
    }
  } catch (error) {
    console.warn('[updates] update check failed:', error)
    return null
  }
}

export function skipUpdateVersion(version: string): void {
  if (parseVersion(version)) store.set(SKIPPED_VERSION_KEY, normalizedVersion(version))
}

File: qt_frontend/src/UpdateChecker.cpp
#include "UpdateChecker.h"

#include <QApplication>
#include <QDateTime>
#include <QDesktopServices>
#include <QDebug>
#include <QJsonDocument>
#include <QJsonObject>
#include <QMessageBox>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QPushButton>
#include <QRegularExpression>
#include <QSettings>
#include <QStringList>
#include <QUrl>

namespace {
constexpr qint64 kCheckIntervalMs = 24LL * 60 * 60 * 1000;
const QUrl kReleaseApi("https://api.github.com/repos/NoGoat/Track-N-Race/releases/latest");
const QUrl kReleasePage("https://github.com/NoGoat/Track-N-Race/releases/latest");

struct Version {
    int core[3] = {};
    QStringList prerelease;
};

QString normalizedVersion(QString value) {
    value = value.trimmed();
    if (value.startsWith('v', Qt::CaseInsensitive) && value.size() > 1 && value[1].isDigit())
        value.remove(0, 1);
    return value;
}

bool parseVersion(const QString& input, Version& out) {
    static const QRegularExpression expression(
        R"(^v?(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$)");
    const auto match = expression.match(input.trimmed());
    if (!match.hasMatch()) return false;
    for (int i = 0; i < 3; ++i) out.core[i] = match.captured(i + 1).toInt();
    out.prerelease = match.captured(4).isEmpty()
        ? QStringList{} : match.captured(4).split('.');
    return true;
}

int compareVersions(const QString& leftText, const QString& rightText, bool& valid) {
    Version left, right;
    valid = parseVersion(leftText, left) && parseVersion(rightText, right);
    if (!valid) return 0;
    for (int i = 0; i < 3; ++i)
        if (left.core[i] != right.core[i]) return left.core[i] > right.core[i] ? 1 : -1;
    if (left.prerelease.isEmpty() != right.prerelease.isEmpty())
        return left.prerelease.isEmpty() ? 1 : -1;
    const int count = qMax(left.prerelease.size(), right.prerelease.size());
    for (int i = 0; i < count; ++i) {
        if (i >= left.prerelease.size()) return -1;
        if (i >= right.prerelease.size()) return 1;
        if (left.prerelease[i] == right.prerelease[i]) continue;
        bool leftNumber = false, rightNumber = false;
        const int leftValue = left.prerelease[i].toInt(&leftNumber);
        const int rightValue = right.prerelease[i].toInt(&rightNumber);
        if (leftNumber && rightNumber) return leftValue > rightValue ? 1 : -1;
        if (leftNumber != rightNumber) return leftNumber ? -1 : 1;
        return QString::compare(left.prerelease[i], right.prerelease[i], Qt::CaseSensitive) > 0 ? 1 : -1;
    }
    return 0;
}
}

UpdateChecker::UpdateChecker(QWidget* dialogParent)
    : QObject(dialogParent), dialogParent_(dialogParent),
      network_(new QNetworkAccessManager(this)) {}

void UpdateChecker::checkOnStartup() {
    QSettings settings("TrackNRace", "NativeRecorder");
    if (!settings.value("updates/enabled", true).toBool()) return;
    const qint64 now = QDateTime::currentMSecsSinceEpoch();
    if (now - settings.value("updates/lastCheckAt", 0).toLongLong() < kCheckIntervalMs) return;
    // Store the attempt first so an outage does not retry on every launch.
    settings.setValue("updates/lastCheckAt", now);

    QNetworkRequest request(kReleaseApi);
    request.setRawHeader("Accept", "application/vnd.github+json");
    request.setRawHeader("User-Agent", "Track-N-Race");
    request.setRawHeader("X-GitHub-Api-Version", "2022-11-28");
    QNetworkReply* reply = network_->get(request);
    connect(reply, &QNetworkReply::finished, this, [this, reply] {
        reply->deleteLater();
        if (reply->error() != QNetworkReply::NoError) {
            qWarning().noquote() << "[updates] update check failed:" << reply->errorString();
            return;
        }
        const QJsonObject body = QJsonDocument::fromJson(reply->readAll()).object();
        if (!body.value("tag_name").isString()) {
            qWarning() << "[updates] latest release did not include a tag name";
            return;
        }
        const QString latest = normalizedVersion(body.value("tag_name").toString());
        const QString current = normalizedVersion(QApplication::applicationVersion());
        bool valid = false;
        if (compareVersions(latest, current, valid) <= 0) {
            if (!valid) qWarning().noquote() << "[updates] could not compare versions" << current << latest;
            return;
        }
        QSettings settings("TrackNRace", "NativeRecorder");
        if (normalizedVersion(settings.value("updates/skippedVersion").toString()) == latest) return;

        QMessageBox dialog(QMessageBox::Information, "Update Available",
            QString("Track N Race %1 is available.").arg(latest), QMessageBox::NoButton,
            dialogParent_);
        dialog.setInformativeText(
            QString("Automatic updates are not supported. Download opens the GitHub Releases page.\n\n"
                    "Installed: %1\nLatest: %2").arg(current, latest));
        QPushButton* skip = dialog.addButton("Skip this version", QMessageBox::DestructiveRole);
        dialog.addButton("Remind me Later", QMessageBox::RejectRole);
        QPushButton* download = dialog.addButton("Download", QMessageBox::AcceptRole);
        dialog.exec();
        if (dialog.clickedButton() == skip)
            settings.setValue("updates/skippedVersion", latest);
        else if (dialog.clickedButton() == download)
            QDesktopServices::openUrl(kReleasePage);
    });
}
